gitlab/resource_gitlab_user_impersonation_token.go

```diff
@@ -0,0 +1,136 @@
+package gitlab
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabUserImpersonationToken() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabUserImpersonationTokenCreate,
+		Read:   resourceGitlabUserImpersonationTokenRead,
+		Delete: resourceGitlabUserImpersonationTokenDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"user": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"active": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+			"revoked": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"token": {
+				Type:      schema.TypeString,
+				Sensitive: true,
+				Computed:  true,
+			},
+			"scopes": {
+				Type:     schema.TypeSet,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.StringInSlice([]string{"api", "read_user"}, true),
+				},
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"expires_at": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabUserImpersonationTokenCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	user := d.Get("user").(int)
+	options := &gitlab.CreateImpersonationTokenOptions{
+		Name: gitlab.String(d.Get("name").(string)),
+	}
+
+	if v, ok := d.GetOk("scopes"); ok {
+		options.Scopes = stringSetToStringSlice(v.(*schema.Set))
+	}
+
+	if v, ok := d.GetOk("expires_at"); ok {
+		layout := "2006-01-02"
+		t, _ := time.Parse(layout, v.(string))
+		options.ExpiresAt = &t
+	}
+
+	log.Printf("[DEBUG] create gitlab user impersonation %s for user %d", *options.Name, user)
+	impersonationToken, _, err := client.Users.CreateImpersonationToken(user, options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%d/%d", user, impersonationToken.ID))
+	// We need to set token here instead of read, as it is only returned once
+	d.Set("token", impersonationToken.Token)
+
+	return resourceGitlabUserImpersonationTokenRead(d, meta)
```

Comment:
I'm curious why you need to call the read function if you already have the data you need from the `impersonationToken` object?

Reply:
Never mind. I see it is a recommended practice in the docs. Carry on!

Reply:
guess I can resolve this ?